lms/djangoapps/commerce/views.py

```diff
@@ -0,0 +1,157 @@
+    authentication_classes = (SessionAuthenticationAllowInactiveUser,)
+    permission_classes = (IsAuthenticated,)
+
+    def _is_data_valid(self, request):
+        """
+        Validates the data posted to the view.
+
+        Arguments
+            request -- HTTP request
+
+        Returns
+            Tuple (data_is_valid, course_key, error_msg)
+        """
+        course_id = request.DATA.get('course_id')
+
+        if not course_id:
+            return False, None, u'Field course_id is missing.'
+
+        try:
+            course_key = CourseKey.from_string(course_id)
+            courses.get_course(course_key)
+        except Exception as ex:  # pylint: disable=broad-except
+            log.exception(u'Unable to locate course matching %s.', course_id)
+            return False, None, ex.message
+
+        return True, course_key, None
+
+    def _get_jwt(self, user):
+        """
+        Returns a JWT object with the specified user's info.
+        """
+        data = {
+            'username': user.username,
+            'email': user.email
+        }
+        return jwt.encode(data, getattr(settings, 'ECOMMERCE_API_SIGNING_KEY', None))
+
+    def _enroll(self, course_key, user):
+        """ Enroll the user in the course. """
+        CourseEnrollment.enroll(user, course_key, True)
+
+    def post(self, request, *args, **kwargs):  # pylint: disable=unused-argument
+        """
+        Attempt to create the order and enroll the user.
+        """
+        user = request.user
+        valid, course_key, error = self._is_data_valid(request)
+        if not valid:
+            return DetailResponse(error, status=HTTP_406_NOT_ACCEPTABLE)
+
+        # Ensure that the E-Commerce API is setup properly
+        ecommerce_api_url = getattr(settings, 'ECOMMERCE_API_URL', None)
+        ecommerce_api_signing_key = getattr(settings, 'ECOMMERCE_API_SIGNING_KEY', None)
+
+        if not (ecommerce_api_url and ecommerce_api_signing_key):
+            self._enroll(course_key, user)
+            msg = u'E-Commerce API not setup. Enrolled {} in {} directly.'.format(user.username, unicode(course_key))
+            log.debug(msg)
+            return DetailResponse(msg)
+
+        # Default to honor mode. In the future we may expand this view to support additional modes.
+        mode = CourseMode.DEFAULT_MODE_SLUG
+        course_modes = CourseMode.objects.filter(course_id=course_key, mode_slug=mode, sku__isnull=False)
+
+        # If there are no course modes with SKUs, enroll the user without contacting the external API.
+        if not course_modes.exists():
+            msg = u'The {enrollment_mode} mode for {course_id} does not have a SKU. Enrolling {username} directly.'
+            msg = msg.format(enrollment_mode=mode, course_id=unicode(course_key), username=user.username)
+            log.debug(msg)
+            self._enroll(course_key, user)
```

Comment:
Given that [`_enroll`](https://github.com/edx/edx-platform/pull/7292/files#diff-3e84d5041cfb763b355bac2d2d063ae9R67) operates directly on the student models, this could result in (extremely) undesired behavior when the user is trying to enroll in a professional ed course. The database call will return no course modes matching the filter, since professional ed courses have no honor modes, let alone modes with SKUs. As a result, the user will be enrolled in the course (as "honor," strangely enough) without having first asked them to pay.

This is similar to a mistake we've made in the past during the authentication process. Instead of touching the student models, I would suggest communicating with the enrollment API in process, which has checks in place to avoid behavior like this. If the attempt to enroll fails, [the JS](https://github.com/edx/edx-platform/commit/39f82ab08b4858c02de1134fced7661191ed3f1f#diff-a2ed964d4369155fbb342cd8ba669e4cR50) is designed to send the user along to the track selection page regardless.